#1: New feature - record all the step

sampraxis opened the pull request:
* New feature: Add step number indicator.
* New feature: Add record in all steps.
* Fix the bugs that doesn't follow Moodle standard.
* Fix security issue.
* Refactor critical parts (such as web service).

sampraxis committed:
#3: Add step number indicator

New feature: Add step number indicator

sampraxis committed:
Update release version

sampraxis committed:
#1: Record all steps

Fix the bugs that doesn't follow Moodle standard

Fix security issue

New feature: Add record in all steps

Snowdle commented on classes/app/ai/attempt/interfaces/mapper.php:
```diff
@@ -0,0 +1,19 @@
+//     * @param iterable<entity>|entity[] $records
+//     * @return collection<entity>|entity[]
+//     */
+//    public function map_to_collection(iterable $records): collection;
```
I don't see a reason to keep these comments here

Snowdle commented on classes/app/ai/attempt/repository.php:
```diff
@@ -0,0 +1,171 @@
+use assignsubmission_pxaiwriter\app\interfaces\factory as base_factory;
+use dml_exception;
+use moodle_database;
+use stdClass;
```
`stdClass` does not seem to be used

Snowdle commented on classes/app/ai/formatter.php:
```diff
@@ -0,0 +1,31 @@
+    private function limit_newlines(string $text): string
+    {
+        $pattern = '((?:\r\n?|\n)+)$|(?:\r\n?|\n){2,}';
+        return preg_replace("#$pattern#", "\n\n", $text);
```
Why not just have the pattern directly in the preg_replace?

Snowdle commented on classes/app/ai/history/archive.php:
```diff
@@ -0,0 +1,122 @@
+            return $old_history;
+        }
+
+        return $this->push_history($history, $ai_text);
```
I think null coalescense would be nicer here. Something like: 
`return $old_history ?? $this->push_history($history, $ai_text);`

sampraxis committed:
#1: Record all the steps

Fix submission overview cannot parse newline to HTML line break

Fix minor issue

Snowdle commented on classes/app/ai/openai/api.php:
```diff
@@ -0,0 +1,116 @@
+
+    private function get_first_choice(array $choices): array
+    {
+        return $choices[0];
```
Using array pointers with `reset()` might be a better option

sampraxis replied:
Just be mindful that ``reset()`` not only return the first element but also reset the iteration pointer of the array.
In this situation there will be one or more elements in the indexed array.

Snowdle commented on classes/app/helper/hash/factory.php:
```diff
@@ -0,0 +1,20 @@
+namespace assignsubmission_pxaiwriter\app\helper\hash;
+
+
+//use assignsubmission_pxaiwriter\app\helper\hash\interfaces\hash_function;
```
Remove this comment

Snowdle commented on classes/app/helper/times/day.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function get_end_of_day(): DateTime
+    {
+        $start_of_day = clone $this->date;
```
Just out of curiosity: why use clone instead of just referencing $this->date? I don't think I've ever seen that in use before :smile:

sampraxis replied:
Because ``$this->date`` is not an immutable ``DateTime`` object and we want a copy of object not the reference of it.
And because of that if you modified something that return by ``get_end_of_day()``
It will modify the ``$this->date`` itself.  Which it is undesire behavior.

You can also doing ``new DateTime($this->date->format('Y-m-d H:i:s'), $this->date->getTimezone())``.

And in this case ``clone`` is a bit easier to read and it copy the object without an extra work :smile: 

There is also ``DateTimeImmutable`` that can also be use in this situation.

Example
```php


$date = new DateTime('2023-02-01 00:00:00');
$a = $date;
$b = $date;
$a->setTime(12, 0, 0);
$b->setTime(9, 0, 0);

echo $a->format('Y-m-d H:i:s'); // 2023-02-01 09:00:00
echo PHP_EOL;
echo $b->format('Y-m-d H:i:s'); // 2023-02-01 09:00:00

$date = new DateTime('2023-02-01 00:00:00');
$a = clone $date;
$b = clone $date;

$a->setTime(12, 0, 0);
$b->setTime(9, 0, 0);

echo PHP_EOL;
echo $a->format('Y-m-d H:i:s'); // 2023-02-01 12:00:00
echo PHP_EOL;
echo $b->format('Y-m-d H:i:s'); // 2023-02-01 09:00:00

```

Snowdle commented on classes/app/helper/times/factory.php:
```diff
@@ -0,0 +1,33 @@
+
+    public function current_timestamp(): int
+    {
+        return time();
```
"timestamp" is a fairly broad term. I think it would be better to rename this function to something like `unix_timestamp()`, to better indicate which output is expected

sampraxis replied:
Call it ``unix_timestamp()`` doesn't not help to solve the naming convention. Because it return the current time in the unix timestamp.
Maybe ``current_time()`` will be enough in this context.

Snowdle commented on classes/app/setting/admin_settings.php:
```diff
@@ -0,0 +1,78 @@
+
+    public function get_url(): string
+    {
+        return $this->config->url ?? 'https://api.openai.com/v1/completions';
```
I think we should throw en exception instead, if no API URL has been configured.

sampraxis replied:
Something moodle doesn't got save the configuration and it leave empty even though it has a default value.
So this is just an insurance for it to return a default value.

If it has no default value, then it will make sense to throw it.

Snowdle commented on externallib.php:
```diff
@@ -33,7 +36,7 @@ public static function do_ai_magic($contextid, $jsondata)
             $attempt_record = self::getAIAttemptRecord($assignmentid, $userid);
 
-            if (self::isExceedingAttemptCount($attempt_record)) {
+            if (false) {
```
This is never being run?

sampraxis replied:
``externallib.php`` is not used by anything. (Might be deprecate in the future as well)
I have remove it in the upcoming commit.

Snowdle commented on lang/en/assignsubmission_pxaiwriter.php:
```diff
@@ -93,4 +93,13 @@
+$string['event_ai_text_generated'] = 'AI writer text generated';
+
+// Errors
+$string['error_course_module_not_found_by_assign_id'] = 'Cannot found course module by assignment id {$a}';
```
Should be "Could not find..."

Snowdle commented on lang/en/assignsubmission_pxaiwriter.php:
```diff
@@ -93,4 +93,13 @@
+
+// Errors
+$string['error_course_module_not_found_by_assign_id'] = 'Cannot found course module by assignment id {$a}';
+$string['error_user_exceed_attempts'] = 'You have already exceed the attempt limit';
```
Should be "You have already exceeded..."

Snowdle reviewed (changes requested):
Looks good, I only have some minor questions and suggestions

sampraxis committed:
#1: Record all the steps

Add the suggestion from #5 by @Snowdle

Fix expand AI text doesn't replace selected text properly

Snowdle commented on lang/en/assignsubmission_pxaiwriter.php:
```diff
@@ -93,4 +93,14 @@
+$string['event_ai_text_generated'] = 'AI writer text generated';
+
+// Errors
+$string['error_course_module_not_found_by_assign_id'] = 'Could not found course module by assignment id {$a}';
```
The word "found" still needs to be changed to "find"

sampraxis committed:
#1: Record all the steps

Add the suggestion from #5 by @Snowdle

Update language string

Snowdle reviewed (approved):
:rocket:
